rbd/admin/task_test.go

```diff
@@ -0,0 +1,423 @@
+			assert.NoError(t, err)
+
+			imageID, err := img.GetId()
+			assert.NoError(t, err)
+
+			assert.NoError(t, img.Trash(0))
+			assert.NoError(t, img.Close())
+
+			tr, err := ta.AddTrashRemove(NewImageSpec(tt.args.pool, tt.args.namespace, imageID))
+			assert.NoError(t, err)
+
+			assert.Equal(t, imageID, tr.Refs.ImageID)
+			assert.Equal(t, defaultPoolName, tr.Refs.PoolName)
+			assert.Equal(t, tt.args.namespace, tr.Refs.PoolNamespace)
+			assert.Equal(t, "trash remove", tr.Refs.Action)
+
+			trashList := []rbd.TrashInfo{}
+			// wait for the image to be deleted
+			for i := 0; i < 30; i++ {
+				trashList, err = rbd.GetTrashList(ioctx)
+				assert.NoError(t, err)
+
+				if len(trashList) == 0 {
+					break
+				}
+				time.Sleep(time.Second)
+			}
+			assert.Equal(t, 0, len(trashList))
+
+		})
+	}
+}
+
+func TestTaskAdmin_AddFlatten(t *testing.T) {
+	parentImageName := "parent"
+
+	ensureDefaultPool(t)
+	conn := getConn(t)
+
+	ioctx, err := conn.OpenIOContext(defaultPoolName)
+	require.NoError(t, err)
+	defer ioctx.Destroy()
+
+	assert.NoError(t, rbd.NamespaceCreate(ioctx, testNS))
+	defer func() {
+		assert.NoError(t, rbd.NamespaceRemove(ioctx, testNS))
+	}()
+
+	ta := getAdmin(t).Task()
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ioctx.SetNamespace(tt.args.namespace)
+
+			options := rbd.NewRbdImageOptions()
+			assert.NoError(t,
+				options.SetUint64(rbd.ImageOptionOrder, uint64(testImageOrder)))
+			err = rbd.CreateImage(ioctx, parentImageName, testImageSize, options)
+			assert.NoError(t, err)
+
+			parentImage, err := rbd.OpenImage(ioctx, parentImageName, rbd.NoSnapshot)
+			assert.NoError(t, err)
+
+			snap, err := parentImage.CreateSnapshot(tt.args.imageName)
+			assert.NoError(t, err)
+			err = snap.Protect()
+			assert.NoError(t, err)
+
+			assert.NoError(t, options.SetUint64(rbd.ImageOptionFormat, uint64(2)))
+
+			assert.NoError(t, rbd.CloneImage(ioctx, parentImageName, tt.args.imageName, ioctx, tt.args.imageName, options))
+
+			childImage, err := rbd.OpenImage(ioctx, tt.args.imageName, rbd.NoSnapshot)
+			assert.NoError(t, err)
+			parentInfo, err := childImage.GetParent()
+			assert.NoError(t, err)
+			assert.Equal(t, parentImageName, parentInfo.Image.ImageName)
+			assert.Equal(t, tt.args.imageName, parentInfo.Snap.SnapName)
+
+			tr, err := ta.AddFlatten(NewImageSpec(tt.args.pool, tt.args.namespace, tt.args.imageName))
+			assert.NoError(t, err)
+
+			assert.Equal(t, tt.args.imageName, tr.Refs.ImageName)
+			assert.Equal(t, defaultPoolName, tr.Refs.PoolName)
+			assert.Equal(t, tt.args.namespace, tr.Refs.PoolNamespace)
+			assert.Equal(t, "flatten", tr.Refs.Action)
+
+			// wait for the image to be flattened
+			for i := 0; i < 30; i++ {
+				_, err = childImage.GetParent()
+				if errors.Is(err, rbd.RbdErrorNotFound) {
+					break
+				}
+				assert.NoError(t, err)
+
+				time.Sleep(time.Second)
+			}
+			assert.Error(t, err)
+
+			assert.NoError(t, childImage.Close())
```

Comment:
I'd prefer  to see things like this as defers up near the lines that assign/create/open the objects. I find that it keeps more local context for the reader as well as eases refactoring in the future.